### ISSUE TITLE
Assignability issue in narrowed constrained key type 

### ISSUE THREAD

**Issue by tinganho** [Microsoft/TypeScript#12635]
The below code:
```ts
class M {
    __m: any;
}
class C {
    __c: any;
}
type D = typeof M | typeof C;
interface R {
    [i: string]: D;
}
function f1(p: D) {
}
interface P {
    p1: string;
    p2: string;
}
function f2<K extends keyof P>(p: K) {
    let r: R = {};
    if (p in r) {
        f1(r[p]); // Error  `R[K]` is not assignable to `D`
    }
}
```
Gives me the error `R[K]` is not assignable to `D`.

Though isn't `R[K]` equal to `D`?

**Comment by mhegazy** [comment]
you do not really need the generic type patamter here. this should be writen as:

```ts
function f2(p: keyof P) {
    let r: R = {};
    if (p in r) {
        f1(r[p]); 
    }
}
```

**Comment by tinganho** [comment]
It seems to work without a generic type parameter. Though, in my actual code I'm using the following signature: 

```ts
public set<K extends keyof P>(prop: K, value: P[K] | null | undefined): void;
```

**Comment by ahejlsberg** [comment]
This no longer reproduces in the nightly build. I believe it was fixed in #12770.
